gnome-image-installer/pages/install/gis-scribe.c

```diff
@@ -1150,6 +1186,128 @@ gis_scribe_begin_verify (GisScribe *self,
   return g_object_ref (gpg_stdin);
 }
 
+static void
+checksum_in_thread (GTask        *task,
+                    gpointer      source,
+                    gpointer      task_data,
+                    GCancellable *cancellable)
+{
+  GisScribe *self = source;
+  GisScribeChecksumData *checksum_data = task_data;
+  gsize len;
+  guint8 buf[BUFFER_SIZE];
+  g_autoptr(GChecksum) sha256sum = g_checksum_new (G_CHECKSUM_SHA256);
+  g_autoptr(GError) error = NULL;
+  const gchar *digest;
+
+  for (;;) {
+    if (!g_input_stream_read_all (checksum_data->input, buf, sizeof (buf), &len,
+                                  NULL, &error))
+      {
+        task_return_error (self, task, g_steal_pointer (&error));
+        return;
+      }
+
+    if (len == 0)
+      break;
+
+    g_checksum_update (sha256sum, buf, len);
+  }
+
+  digest = g_checksum_get_string (sha256sum);
+  if (g_strcmp0 (digest, checksum_data->expected_checksum) != 0)
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("Image checksum ‘%s‘ does not match expected checksum ‘%s‘."),
+          digest, checksum_data->expected_checksum);
+      return;
+    }
+
+  g_task_return_boolean (task, TRUE);
+}
+
+static GOutputStream *
+gis_scribe_begin_verify_checksum (GisScribe           *self,
+                                  GCancellable        *cancellable,
+                                  GAsyncReadyCallback  callback,
+                                  gpointer             data)
+{
+  g_autoptr(GTask) task = g_task_new (self, cancellable, callback, data);
+  GisScribeChecksumData *task_data = g_slice_new0 (GisScribeChecksumData);
+  g_autofree gchar *checksum_path = g_file_get_path (self->checksum);
+  g_autofree gchar *checksum_contents = NULL;
+  gsize checksum_len;
+  gchar *cur;
+  gint pipefd[2];
+  g_autoptr(GError) error = NULL;
+
+  g_task_set_source_tag (task, GUINT_TO_POINTER (GIS_SCRIBE_TASK_VERIFY));
+  g_task_set_task_data (task, task_data,
+                        (GDestroyNotify) gis_scribe_checksum_data_free);
+
+  if (!g_file_query_exists (self->checksum, NULL))
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("The checksum file ‘%s’ does not exist."),
+          checksum_path);
+      return NULL;
+    }
+
+  /* Read in the checksum file to get the expected checksum. */
+  if (!g_file_load_contents (self->checksum, cancellable, &checksum_contents,
+                             &checksum_len, NULL, &error))
+    {
+      task_return_error (self, task, g_steal_pointer (&error));
+      return NULL;
+    }
+
+  if (checksum_len < CHECKSUM_STRLEN)
+    {
+      task_return_new_error (
+          self, task, GIS_IMAGE_ERROR, GIS_IMAGE_ERROR_VERIFICATION_FAILED,
+          _("The checksum file ‘%s’ does not contain %d characters."),
+          checksum_path, CHECKSUM_STRLEN);
+      return NULL;
+    }
```

Comment:
Seems redundant with the block below?

Reply:
Right. I had something different originally but this is unnecessary.